### PR TITLE
Optimize pipeline task run query

### PR DESCRIPTION
**Pull request by connorwstein**
At the scale of the load test (15k task runs) the speed up is (1.859594234s/8.343933ms) = ~222x. Not totally sure why the local explain analyze is waay slower for the old query vs running the load test but here are the two explains:

*Before*
```
| Limit  (cost=1646.30..1647.01 rows=1 width=130) (actual time=70185.760..70185.925 rows=1 loops=1)
|   ->  Nested Loop  (cost=1646.30..3425.83 rows=2513 width=130) (actual time=70185.747..70185.895 rows=1 loops=1)
|         ->  HashAggregate  (cost=1646.01..1708.51 rows=2513 width=8) (actual time=70185.687..70185.811 rows=1 loops=1)
|               Group Key: pipeline_task_runs_1.id
|               Filter: (bool_and((predecessor_unfinished_runs.finished_at IS NOT NULL)) OR (count(predecessor_unfinished_runs.id) = 0))
|               ->  Hash Right Join  (cost=416.24..1441.47 rows=27273 width=24) (actual time=139.404..34446.488 rows=5000000 loops=1)
|                     Hash Cond: ((predecessor_unfinished_runs.pipeline_task_spec_id = predecessor_specs.id) AND (predecessor_unfinished_
|                     ->  Seq Scan on pipeline_task_runs predecessor_unfinished_runs  (cost=0.00..340.00 rows=15000 width=28) (actual tim
|                     ->  Hash  (cost=341.24..341.24 rows=5000 width=20) (actual time=139.314..139.403 rows=5000 loops=1)
|                           Buckets: 8192  Batches: 1  Memory Usage: 318kB
|                           ->  Hash Join  (cost=46.49..341.24 rows=5000 width=20) (actual time=0.736..104.121 rows=5000 loops=1)
|                                 Hash Cond: (pipeline_task_runs_1.pipeline_task_spec_id = pipeline_task_specs.id)
|                                 ->  Index Scan using idx_pipeline_task_runs_unfinished on pipeline_task_runs pipeline_task_runs_1  (cos
|                                 ->  Hash  (cost=39.33..39.33 rows=550 width=8) (actual time=0.678..0.732 rows=15 loops=1)
|                                       Buckets: 1024  Batches: 1  Memory Usage: 9kB
|                                       ->  Hash Right Join  (cost=22.38..39.33 rows=550 width=8) (actual time=0.315..0.612 rows=15 loops
|                                             Hash Cond: (predecessor_specs.successor_id = pipeline_task_specs.id)
|                                             ->  Seq Scan on pipeline_task_specs predecessor_specs  (cost=0.00..15.50 rows=550 width=8)
|                                             ->  Hash  (cost=15.50..15.50 rows=550 width=4) (actual time=0.266..0.284 rows=15 loops=1)
|                                                   Buckets: 1024  Batches: 1  Memory Usage: 9kB
|                                                   ->  Seq Scan on pipeline_task_specs  (cost=0.00..15.50 rows=550 width=4) (actual time
|         ->  Index Scan using pipeline_task_runs_pkey on pipeline_task_runs  (cost=0.29..0.67 rows=1 width=130) (actual time=0.039..0.04
|               Index Cond: (id = pipeline_task_runs_1.id)
| Planning Time: 0.533 ms
| Execution Time: 70186.177 ms
```

*After*
```
 | QUERY PLAN
|----------------------------------------------------------------------------------------------------------------------------------------
| Limit  (cost=0.86..8.12 rows=1 width=130) (actual time=0.273..0.382 rows=1 loops=1)
|   ->  Merge Join  (cost=0.86..18269.00 rows=2513 width=130) (actual time=0.256..0.342 rows=1 loops=1)
|         Merge Cond: (pipeline_task_runs.id = runs.id)
|         ->  Index Scan using pipeline_task_runs_pkey on pipeline_task_runs  (cost=0.29..590.28 rows=15000 width=130) (actual time=0.022
|         ->  GroupAggregate  (cost=0.57..17584.67 rows=2513 width=8) (actual time=0.175..0.229 rows=1 loops=1)
|               Group Key: runs.id
|               Filter: (bool_and((preds.finished_at IS NOT NULL)) OR (count(preds.*) = 0))
|               ->  Nested Loop Left Join  (cost=0.57..17147.28 rows=49985 width=170) (actual time=0.082..0.185 rows=2 loops=1)
|                     ->  Index Scan using pipeline_task_runs_pkey on pipeline_task_runs runs  (cost=0.29..590.28 rows=5000 width=33) (ac
|                           Filter: (finished_at IS NULL)
|                           Rows Removed by Filter: 4
|                     ->  Index Scan using pipeline_task_runs_pkey on pipeline_task_runs preds  (cost=0.29..3.21 rows=10 width=170) (actu
|                           Index Cond: (id = ANY (runs.predecessor_task_run_ids))
| Planning Time: 0.906 ms
| Execution Time: 0.576 ms
```

### CODE REVIEW

File: core/services/pipeline/orm_test.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+	gormv1 "github.com/jinzhu/gorm"
+	p "github.com/smartcontractkit/chainlink/core/services/postgres"
+	"github.com/smartcontractkit/chainlink/core/store/migrations"
+	"github.com/stretchr/testify/require"
+	"gopkg.in/guregu/null.v4"
+	"gorm.io/driver/postgres"
+	"gorm.io/gorm"
+)
+
+func TestClaimTaskScale(t *testing.T) {
```
Comment (connorwstein) [inline core/services/pipeline/orm_test.go:18]: I think we could leave this in but just t.Skip it?

File: core/services/pipeline/orm.go [hunk 2]
```diff
@@ -159,6 +157,47 @@ func (o *orm) CreateRun(ctx context.Context, jobID int32, meta map[string]interf
...
+		var preds = make(map[int32][]int32)
+		for _, taskSpec := range ts {
+			if taskSpec.SuccessorID.Valid {
+				preds[int32(taskSpec.SuccessorID.Int64)] = append(preds[int32(taskSpec.SuccessorID.Int64)], taskSpec.ID)
+			}
+		}
+		for _, taskRun := range tr {
+			if preds, ok := preds[taskRun.PipelineTaskSpecID]; ok {
+				// Find the corresponding taskRun for each pred
+				// TODO: clever optimize?
```
Comment (connorwstein) [inline core/services/pipeline/orm.go:186]: todo spend a little more time on this - its sort of like a graph cloning problem

File: core/services/pipeline/orm.go [hunk 3]
```diff
@@ -159,6 +157,47 @@ func (o *orm) CreateRun(ctx context.Context, jobID int32, meta map[string]interf
             SELECT ? AS pipeline_run_id, id AS pipeline_task_spec_id, type, index, NOW() AS created_at
             FROM pipeline_task_specs
             WHERE pipeline_spec_id = ?`, run.ID, run.PipelineSpecID).Error
+		if err != nil {
+			return err
+		}
+
+		// Ammend the TaskRuns with their predecessors.
+		var tr []TaskRun
```
Comment (samsondav) [inline core/services/pipeline/orm.go:165]: Can we save a query by using `RETURNING` in the INSERT and scanning into the TaskRun array?

File: core/services/pipeline/orm.go (hunk 2 above)
Comment (samsondav) [reply core/services/pipeline/orm.go:186]: Can we pull this out into a neatly named function?

File: core/services/pipeline/orm.go [hunk 4]
```diff
@@ -209,22 +248,16 @@ func (o *orm) processNextUnclaimedTaskRun(ctx context.Context, fn ProcessTaskRun
...
+			SELECT * FROM pipeline_task_runs WHERE id IN (
+				SELECT runs.id FROM pipeline_task_runs AS runs
+				LEFT JOIN pipeline_task_runs AS preds ON preds.id = ANY (runs.predecessor_task_run_ids)
+				WHERE runs.finished_at is null
+				GROUP BY runs.id
+				HAVING (bool_and(preds.finished_at is not null) 
+					OR count(preds) = 0)
+			)
+           LIMIT 1
+           FOR UPDATE OF pipeline_task_runs SKIP LOCKED;
```
Comment (samsondav) [inline core/services/pipeline/orm.go:260]: This is much better. I'm kicking myself for not thinking of it before, I must have spent so much time staring at this query I forgot the obvious.

File: core/services/pipeline/orm_test.go (hunk 1 above)
Comment (samsondav) [reply core/services/pipeline/orm_test.go:18]: IIRC golang actually has a performance testing mode? Can it fall under that?

File: go.mod [hunk 5]
```diff
@@ -60,5 +60,6 @@ require (
 	gonum.org/v1/gonum v0.8.2
 	gopkg.in/gormigrate.v1 v1.6.0
 	gopkg.in/guregu/null.v4 v4.0.0
+	gorm.io/driver/postgres v1.0.6
```
Comment (samsondav) [inline go.mod:63]: Where is this used?

File: core/store/migrations/migration1611174407/migrate.go [hunk 6]
```diff
@@ -0,0 +1,12 @@
+package migration1611174407
+
+import "github.com/jinzhu/gorm"
+
+// Migrate makes the explicit the pre-existing
+// implicit assumption that lowercase external initiator names are unique
+func Migrate(tx *gorm.DB) error {
+	return tx.Exec(`
+		ALTER TABLE pipeline_task_runs 
+		ADD COLUMN predecessor_task_run_ids integer[];
```
Comment (samsondav) [inline core/store/migrations/migration1611174407/migrate.go:10]: I wonder if we want to make it non-nullable to empty arrays have to be explicit?

**Review by samsondav — CHANGES_REQUESTED** [review]

File: core/services/pipeline/orm.go (hunk 2 above)
Comment (samsondav) [reply core/services/pipeline/orm.go:186]: Also I have a working optimised path for doing something like this, you might be able to use it:

https://github.com/smartcontractkit/chainlink/commit/7a85aeeb11c45c2d583caa281177c6eedf22486d#diff-93a69845888e9fa37d8e5ec7d17ff8f7003b52f598cf0367b1944ed1466966eaR205

Check the whole diff here, there might also be some tests you can use:

https://github.com/smartcontractkit/chainlink/compare/chore/176526416-optimize-work-queue?expand=1

File: go.mod (hunk 5 above)
Comment (connorwstein) [reply go.mod:63]: used in the scale test

File: core/services/pipeline/orm_test.go (hunk 1 above)
Comment (samsondav) [reply core/services/pipeline/orm_test.go:18]: Sorry its called benchmark mode

Comment (connorwstein) [reply core/services/pipeline/orm_test.go:18]: oh I think you just mean the benchmarking? yeah I was thinking about doing that but wanted to still use the testify stuff. Seems easy enough to just write my own requireNoError function though so ill go ahead and convert it to a benchmark

File: core/services/pipeline/orm.go (hunk 2 above)
Comment (connorwstein) [reply core/services/pipeline/orm.go]: I think they are slightly different problems, but it was fairly simple to optimize so I just optimized it and pulled out a helper/unit test.

File: core/services/pipeline/orm.go [hunk 7]
```diff
@@ -154,11 +188,35 @@ func (o *orm) CreateRun(ctx context.Context, jobID int32, meta map[string]interf
 		// Create the task runs
 		err = tx.Exec(`
             INSERT INTO pipeline_task_runs (
-            	pipeline_run_id, pipeline_task_spec_id, type, index, created_at
+            	pipeline_run_id, pipeline_task_spec_id, type, index, created_at, predecessor_task_run_ids
             )
-            SELECT ? AS pipeline_run_id, id AS pipeline_task_spec_id, type, index, NOW() AS created_at
+            SELECT ? AS pipeline_run_id, id AS pipeline_task_spec_id, type, index, NOW() AS created_at, '{}'
```
Comment (samsondav) [inline core/services/pipeline/orm.go:193]:
```
var runs TaskRuns
err = tx.Raw(`
INSERT INTO pipeline_task_runs (
    pipeline_run_id, pipeline_task_spec_id, type, index, created_at, predecessor_task_run_ids
)
SELECT ? AS pipeline_run_id, id AS pipeline_task_spec_id, type, index, NOW() AS created_at, '{}'
RETURNING pipeline_task_runs.*`).Scan(&runs).Error

```

File: core/services/pipeline/orm.go [hunk 8]
```diff
@@ -152,13 +186,32 @@ func (o *orm) CreateRun(ctx context.Context, jobID int32, meta map[string]interf
...
             FROM pipeline_task_specs
-            WHERE pipeline_spec_id = ?`, run.ID, run.PipelineSpecID).Error
+            WHERE pipeline_spec_id = ?
+			RETURNING *`, run.ID, run.PipelineSpecID).Scan(&trs).Error
```
Comment (samsondav) [inline core/services/pipeline/orm.go:197]: whitespace is weird here

File: core/services/pipeline/orm_test.go [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+}
+
+func requireNoError(b *testing.B, err error) {
+	if err != nil {
+		b.FailNow()
+	}
+}
+
+func BenchmarkClaimTaskQuery(b *testing.B) {
+	//b.Skip()
```
Comment (samsondav) [inline core/services/pipeline/orm_test.go:69]: ✂️  this

File: core/services/pipeline/orm_test.go [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+	assert.Equal(t, pq.Int64Array([]int64{3}), r[3].PredecessorTaskRunIds)
+}
+
+func requireNoError(b *testing.B, err error) {
+	if err != nil {
+		b.FailNow()
+	}
+}
+
+func BenchmarkClaimTaskQuery(b *testing.B) {
```
Comment (samsondav) [inline core/services/pipeline/orm_test.go:68]: How to run out of interest?

**Review by samsondav — DISMISSED** [review]

File: core/services/pipeline/orm_test.go (hunk 10 above)
Comment (connorwstein) [reply core/services/pipeline/orm_test.go:68]: -bench

**Review by spooktheducks — APPROVED** [review]

**Comment by connorwstein** [comment]
Closing in favor of https://github.com/smartcontractkit/chainlink/pull/3882

**Pull request closed without merging**